numpy/core/_asarray.py

```diff
@@ -105,6 +122,16 @@ def asanyarray(a, dtype=None, order=None):
         'A' (any) means 'F' if `a` is Fortran contiguous, 'C' otherwise
         'K' (keep) preserve input order
         Defaults to 'C'.
+    like : array_like
+        Reference array-like object created by a downstream library, such
+        as CuPy, Dask, xarray, etc. If specified, the function will be
+        dispatched to the downstream library, provided it supports the
+        __array_function__ protocol. This will bypass the NumPy's
+        implementation and be computed by the downstream library, the
+        result is thus returned by that library, which may be of the
+        same type as the array provided here, or any other type.
+
+        .. versionadded:: 1.20.0
```

Comment:
I'd suggest templating this as something like

```suggestion
    ${NEP-35_like_arg}

        .. versionadded:: 1.20.0
```
Remember that the version should not be part of the template, in case we add it to more functions in later versions

I have no opinions on what the templated name should be, it doesn't really matter.

Reply:
I was looking at that right now and I'm still not quite certain how to do that, it seems that the only place in NumPy we have this is in https://github.com/numpy/numpy/blob/961b56fb58ee6482cffdce6f61cb679eba654daf/numpy/core/code_generators/ufunc_docstrings.py#L19-L65

There's a whole mechanism to do the substitution. Are you suggesting we do some mechanism like that or is there a simpler way of doing that which I'm not aware of?

Reply:
I suppose you would have to add a line `func.__doc__ = func.__doc__.format()` or so?  That could be part of a decorator maybe... I don't think we have a generic machinery for such substitutions right now. Could be nice I guess, I would not be surprised if there are hundreds places scatter around NumPy where such duplicated docstrings exists...

Reply:
I was just suggesting something dumb like `func.__doc__ = func.__doc__.replace("Some sufficiently unique string", the_full_docstring_for_like)`

Reply:
Any ideas where should `the_full_docstring_for_like` in the example above live? It will be used by files in `numpy/core/` and `numpy/lib/` .

Reply:
Honestly, dunno, maybe `core/overrides.py` although that only makes sense as long as its only used for `like=`, but thats what we have right now.

I would like to wrap this up soon, I did not have a last pass myself, but I think except documenting the experimental nature we are reaching a good state, its good enough that small fixups should be painless later, if necessary.

@pentschev what is the state of the slow-down, did you have any idea on that? What was not clear to me for your solution, was that it seems to me the array-like author would have to override `np. _asarray_with_like ` and not `np.asarray`, which does not seem nice to me. But likely I missed something?

Reply:
I've placed this in `numpy/core/overrides.py` for the time being, as seen in https://github.com/numpy/numpy/pull/16935/commits/1ade021980d5fc07f09a4c2003ea4b8d73cfc416 . For the C docstrings defined in `_add_newdoc`, I added the `.replace(...)` right after the string as it feels very non-intrusive. For the pure Python functions, I used the `func.__doc__ = func._doc___.replace(...)` at the end of each file, as it feels cleaner by avoiding that `replace` call to confuse the reader of a function's implementation, I left it at the end of each file as it's also easier to find them all at one place rather than randomly after each function. If you have a different idea, please let me know.

Reply:
Sorry @seberg , GH didn't update the page so I missed your reply before, but I think we're in agreement.

> @pentschev what is the state of the slow-down, did you have any idea on that? What was not clear to me for your solution, was that it seems to me the array-like author would have to override `np. _asarray_with_like ` and not `np.asarray`, which does not seem nice to me. But likely I missed something?

That's a good reminder, I thought I had already addressed this but clearly forgot. And yes, you're absolutely right, that change was only a simple example to show what I had in mind and is incomplete, we'll need to overwrite the function name, much like as what we do now in the C code when passing the function's name to `array_implement_c_array_function_creation`. I'll work on that.